### PR TITLE
refactor(plugin): v3.0.0

### LINKED ISSUE #314: Upgrade webpack-dev-middleware dependency
**I'm submitting a bug report**

**Webpack version:**
4.0.1

**Webpack Karma version**:
2.0.7

**Karma version**:
2.0.0

**Please tell us about your environment:**
OSX 10.13.3

**Browser:** Chrome 64, Safari 11.0.3

**Current behavior:**
Currently karma-webpack has `webpack-dev-middleware` 1.12.0 listed as a dependency.  This is two major versions behind the current release (3.0.0).

Despite #301 having added webpack v4.x as a `peerDependency`, installing this alongside webpack 4.0.1 will still display warnings because `webpack-dev-middleware` does _not_ list webpack 4.x as a valid peer.  `yarn check` will actually throw an error because of this.

This is the primary roadblock for my project to upgrade to webpack v4.  Our CI will not accept anything for which `yarn check` produces an error.

**Expected/desired behavior:**

karma-webpack should use `webpack-dev-middleware` v2.0.6 which has compatibility with webpack v2.2-4.0, or release a new major version that uses `webpack-dev-middleware` v3.0.0 and drops compatibility with webpack 2 and 3.

* **What is the motivation / use case for changing the behavior?**

wanting to use webpack 4 without warnings in my console when installing packages

### PR DESCRIPTION
**Pull request by rynclark**
**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

Doesn't support webpack 4's new plugin system

**What is the new behavior?**

It supports webpack 4's new plugin system

**Does this PR introduce a breaking change?**
- [x] Yes
- [ ] No

If this PR contains a breaking change, please describe the following...

* Impact: upgrades `webpack-dev-middleware` to latest, making node minimum version `6.11.5`
* Migration path for existing applications: update to webpack 4

Fixes #314

### CODE REVIEW

**Comment by iclanton** [comment]
Any word on when this is going to go in and 3.0.0 is going to get released?

**Comment by limonte** [comment]
FYI, `package-lock.json` is a joke, do not include it in PRs without strong reasons for that:

- https://stackoverflow.com/q/45022048/1331425
- https://github.com/npm/npm/issues/17979

`yarn.lock` works just like expected though.

**Comment by a-tarasyuk** [comment]
+1 to question from @iclanton. Are there any issues that block review/merge? Or maybe need help to fix issues in order to deploy this update.

**Comment by michael-ciniawsky** [comment]
I will release this under the `@next` tag on `npm` once https://status.npmjs.org/incidents/gbk5bjx1169j is resolved so folks can tried it out :)

**Comment by michael-ciniawsky** [comment]
@rynclark Could you do me a favor and split the commits this time with the following commit messages for the `CHANGELOG` 🙃 

```
chore(package): update `webpack-dev-middleware` v1.12.0...3.0.1

BREAKING CHANGE: requires `node >= v6.9.0`
```


```
refactor(karma-webpack): upgrade plugin system (`tapable`)

BREAKING CHANGE: requires `webpack >= v4.0.0`
```

File: package.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
   "description": "Use webpack with karma",
   "license": "MIT",
   "engines": {
-    "node": ">= 4"
+    "node": ">=6.11.5"
```
Comment (michael-ciniawsky) [inline package.json]: `>= 6.9.0`

File: src/karma-webpack.js [hunk 2]
```diff
@@ -91,23 +100,56 @@ function Plugin(
...
-        callback()
+  if (compiler.hooks) {
+    hooks = [
+      {method: 'sync', name: 'invalid'},
+      {method: 'async', name: 'watchRun'},
+      {method: 'async', name: 'run'}
+    ]
+  }
+
+  hooks.forEach(function(hook) {
```
Comment (michael-ciniawsky) [inline src/karma-webpack.js]: `function` is needed here to keep the context ? :)

Comment (a-tarasyuk) [reply src/karma-webpack.js]: I think yes, as there is the second argument in line 150 which set `this` for the callback. The plugin uses ES5 syntax, not ES6, so I think it is ok to use the function here.

**Comment by a-tarasyuk** [comment]
@michael-ciniawsky 
> I will release this under the @next tag

Are there changes which need to be done in order to release stable version?

**Comment by limonte** [comment]
> Are there changes which need to be done in order to release stable version?

https://github.com/webpack-contrib/karma-webpack/milestone/1

**Comment by a-tarasyuk** [comment]
@limonte thx. Is there a change that webpack-4 support will be released early? There are some issues which can take a lot of time, like issues related to performance.

**Comment by michael-ciniawsky** [comment]
Released in [`v4.0.0-beta.0`](https://github.com/webpack-contrib/karma-webpack/releases/tag/v4.0.0-beta.0) 🎉  Thx

**Comment by thijstriemstra** [comment]
@michael-ciniawsky can you give any ETA for karma-webpack 4.0.0?

**Pull request merged**